### PR TITLE
Fix Servlet instrumentation for servlet-api 3.1+

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Fixes https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1083

* Separate Servlet 3.0 and Servlet 3.1 instrumentations
* Separate Jetty 8.0 (depends on Servlet 3.0) and Jetty 9.1 (depends on Servlet 3.1) instrumentations

### CODE REVIEW

File: instrumentation/jetty/jetty-9.1/jetty-9.1.gradle [hunk 1]
```diff
@@ -4,22 +4,21 @@ muzzle {
...
-  library group: 'org.eclipse.jetty', name: 'jetty-server', version: '8.0.0.v20110901'
-  implementation project(':instrumentation:servlet:servlet-3.0')
+  implementation project(':instrumentation:servlet:servlet-3:servlet-3-common')
+  implementation project(':instrumentation:servlet:servlet-3:servlet-3.1')
 
-  // Don't want to conflict with jetty from the test server.
-  testImplementation(project(':testing-common')) {
-    exclude group: 'org.eclipse.jetty', module: 'jetty-server'
-  }
+  // 9.1 is the first Jetty version that introduces Servlet 3.1
```
Comment (iNikem) [inline instrumentation/jetty/jetty-9.1/jetty-9.1.gradle:15]: "supports", not "introduces" :)

File: instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerAdvice.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import io.opentelemetry.instrumentation.auto.servlet.v3.TagSettingAsyncListener;
+import io.opentelemetry.instrumentation.auto.servlet.v3_1.Servlet31CountingHttpServletResponse;
+import io.opentelemetry.trace.Span;
+import java.lang.reflect.Method;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import net.bytebuddy.asm.Advice;
+
+public class Jetty91HandlerAdvice {
```
Comment (iNikem) [inline instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerAdvice.java:34]: This is now identical to `Jetty8HandlerAdvice`, right? Can it be somehow deduplicated?

File: instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/package-info.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ * This module provides support for creating server spans for Jetty Handlers.
+ *
+ * <p>It is possible to write web application running on <a
+ * href="https://www.eclipse.org/jetty/">Eclipse Jetty</a> without actually writing any servlets.
+ * Instead one can use <a
+ * href="https://www.eclipse.org/jetty/documentation//current/jetty-handlers.html">Jetty
+ * Handlers</a>.
+ *
+ * <p>As instrumentation points differ between servlet instrumentations and this one, this module
+ * has its own {@code JettyHandlerInstrumentation} and {@code JettyHandlerAdvice}. But this is the
```
Comment (iNikem) [inline instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/package-info.java:11]: `JettyHandlerAdvice` -> `Jetty91HandlerAdvice`

File: instrumentation/jetty/jetty-testing/src/main/groovy/JettyContinuationHandlerTest.groovy [hunk 4]
```diff
@@ -45,7 +45,7 @@ abstract class JettyContinuationHandlerTest extends JettyHandlerTest {
           continuation.resume()
         }
       } else {
-        handleRequest(baseRequest, response)
+        JettyHandlerTest.handleRequest(baseRequest, response)
```
Comment (iNikem) [inline instrumentation/jetty/jetty-testing/src/main/groovy/JettyContinuationHandlerTest.groovy:48]: Why this change is needed?

File: instrumentation/jsp-2.3/jsp-2.3.gradle [hunk 5]
```diff
@@ -15,7 +15,9 @@ dependencies {
   compileOnly group: 'javax.servlet.jsp', name: 'javax.servlet.jsp-api', version: '2.3.0'
   compileOnly group: 'javax.servlet', name: 'javax.servlet-api', version: '3.1.0'
 
-  testImplementation project(':instrumentation:servlet:servlet-3.0')
+  testImplementation project(':instrumentation:servlet:servlet-3:servlet-3.0')
+  testImplementation project(':instrumentation:servlet:servlet-3:servlet-3.1')
```
Comment (iNikem) [inline instrumentation/jsp-2.3/jsp-2.3.gradle:19]: Why both are needed here?

File: instrumentation/jetty/jetty-9.1/jetty-9.1.gradle (hunk 1 above)
Comment (mateuszrzeszutek) [reply instrumentation/jetty/jetty-9.1/jetty-9.1.gradle]: Fixed

File: instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/package-info.java (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/package-info.java]: Missed that - thanks!

File: instrumentation/jetty/jetty-testing/src/main/groovy/JettyContinuationHandlerTest.groovy (hunk 4 above)
Comment (mateuszrzeszutek) [reply instrumentation/jetty/jetty-testing/src/main/groovy/JettyContinuationHandlerTest.groovy]: It's not -- IntelliJ introduced it on its own when I was moving the file around and I must have missed it. Fixed.

File: instrumentation/jsp-2.3/jsp-2.3.gradle (hunk 5 above)
Comment (mateuszrzeszutek) [reply instrumentation/jsp-2.3/jsp-2.3.gradle:19]: Tomcat 7.X uses servlet 3.0, Tomcat 9.X uses 4.0 (I think). I need to include both instrumentation so that both `test` and `test -PtestLatestDeps=true` pass.

File: instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerAdvice.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerAdvice.java:34]: It's almost identical -- the only differences are `TRACER` and the response wrapper type. Since it's an advice class everything must be static, no dependency injection, no passing parameters -- I don't really know how to parametrize it. Maybe someone here will have a good idea, I didn't, so I just copied the least code that I absolutely had to to make it work.

By the way, servlet 3.0 vs 3.1 has a very similar situation: the only difference is the response wrapper type.

**Comment by iNikem** [comment]
@mateuszrzeszutek  Can you please rebase this PR?

File: instrumentation/jetty/jetty-8.0/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty8HandlerInstrumentation.java [hunk 6]
```diff
@@ -54,16 +56,19 @@ public JettyHandlerInstrumentation() {
...
       "io.opentelemetry.instrumentation.servlet.HttpServletRequestGetter",
       "io.opentelemetry.instrumentation.servlet.ServletHttpServerTracer",
-      "io.opentelemetry.instrumentation.auto.servlet.v3_0.Servlet3HttpServerTracer",
-      "io.opentelemetry.instrumentation.auto.servlet.v3_0.TagSettingAsyncListener",
-      "io.opentelemetry.instrumentation.auto.servlet.v3_0.CountingHttpServletRequest",
-      "io.opentelemetry.instrumentation.auto.servlet.v3_0.CountingHttpServletResponse",
-      "io.opentelemetry.instrumentation.auto.servlet.v3_0.CountingHttpServletResponse$CountingServletOutputStream",
-      "io.opentelemetry.instrumentation.auto.servlet.v3_0.CountingHttpServletResponse$CountingPrintWriter",
-      packageName + ".JettyHttpServerTracer",
+      commonPackageName + ".CountingHttpServletRequest",
```
Comment (anuraaga) [inline instrumentation/jetty/jetty-8.0/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty8HandlerInstrumentation.java:64]: @trask suggested we keep the full names so IDE can refactor them properly. What do you think?

File: instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerInstrumentation.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+@AutoService(Instrumenter.class)
+public final class Jetty91HandlerInstrumentation extends Instrumenter.Default {
+
+  public Jetty91HandlerInstrumentation() {
+    super("jetty", "jetty-9.1");
+  }
+
+  @Override
+  public ElementMatcher<ClassLoader> classLoaderMatcher() {
+    // Jetty 9.0 uses servlet-api 3.1
```
Comment (anuraaga) [inline instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerInstrumentation.java:43]:
```suggestion
    // Jetty 9.1 uses servlet-api 3.1
```

I think

File: instrumentation/servlet/servlet-3/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/Servlet30CountingHttpServletResponse.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    return outputStream;
+  }
+
+  @Override
+  protected int getOutputStreamCounter() {
+    return outputStream == null ? 0 : outputStream.counter;
+  }
+
+  static class CountingServletOutputStream extends ServletOutputStream {
```
Comment (anuraaga) [inline instrumentation/servlet/servlet-3/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/Servlet30CountingHttpServletResponse.java:49]: Can this class be shared in servlet-common? It seems to only depend on servlet (technically only JDK)

**Review by anuraaga — COMMENTED** [review]
Can you start a section in the README about the differences in instrumentation of the servlet 3.0 versions you've made here?

https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/master/instrumentation/servlet

We should fill it in with all the differences and reference information (e.g., jetty versions), but for now just 3.x would be nice. Especially since most of the code looks so similar, it's really hard to see what's changed between versions. Ideally, we can cut down on duplication where possible, but if it's difficult, it's ok but I think we need some text to point people at the places that matter.

**Comment by anuraaga** [comment]
Also really sorry to bring this up so late, I reread the issue #1083 - is it possible to just keep one instrumentation but raise the `compileOnly` dependency to include servlet 3.1? I think it's supposed to be safe for our class to define methods that may not be available at runtime if they're not called. An example of where we do this is Lettuce, we don't have special instrumentation for lettuce 5.2 but support its new API.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/lettuce/lettuce-5.1/src/main/java/io/opentelemetry/instrumentation/auto/lettuce/v5_1/OpenTelemetryTracing.java#L61

I don't know if it works here but wanted to point this out.

File: instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerInstrumentation.java (hunk 7 above)
Comment (mateuszrzeszutek) [reply instrumentation/jetty/jetty-9.1/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty91HandlerInstrumentation.java]: 👍 , missed that one

File: instrumentation/servlet/servlet-3/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/Servlet30CountingHttpServletResponse.java (hunk 8 above)
Comment (mateuszrzeszutek) [reply instrumentation/servlet/servlet-3/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/Servlet30CountingHttpServletResponse.java:49]: Actually this class is the reason why I had to split 3.0 and 3.1 instrumentations 😄 

3.1 introduced two new abstract methods and 1 new type reference (`WriteListener`) so I had to create 2 different `ServletOutputStream` implementations.

**Comment by mateuszrzeszutek** [comment]
> is it possible to just keep one instrumentation but raise the compileOnly dependency to include servlet 3.1? I think it's supposed to be safe for our class to define methods that may not be available at runtime if they're not called. An example of where we do this is Lettuce, we don't have special instrumentation for lettuce 5.2 but support its new API.

Hmm, I'm not sure how it'll work since one of those methods (`setWriteListener`) references a type added in 3.1, but I'll test it -- if the instrumentation works and muzzle check passes then it would be a great way to reduce code duplication.

File: instrumentation/jetty/jetty-8.0/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty8HandlerInstrumentation.java (hunk 6 above)
Comment (mateuszrzeszutek) [reply instrumentation/jetty/jetty-8.0/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty8HandlerInstrumentation.java:64]: Good point, I haven't thought about that.

**Comment by mateuszrzeszutek** [comment]
@anuraaga I've prepared another PR based on your suggestion: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1149

I must say that despite it being a tad hacky I like the new approach more -- no need to split modules or copy-paste entire advice classes.

File: instrumentation/jetty/jetty-8.0/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty8HandlerInstrumentation.java (hunk 6 above)
Comment (trask) [reply instrumentation/jetty/jetty-8.0/src/main/java/io/opentelemetry/instrumentation/auto/jetty/Jetty8HandlerInstrumentation.java:64]: just fyi: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/890#issuecomment-673754096
